packages/repository/src/define-repository-class.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Constructor} from '@loopback/context';
+import assert from 'assert';
+import {Entity, Model} from './model';
+import {
+  CrudRepository,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  EntityCrudRepository,
+  juggler,
+  KeyValueRepository,
+} from './repositories';
+
+/**
+ * Signature for CrudRepository classes
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export interface CrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+> {
+  new (ds: juggler.DataSource): R;
+}
+
+/**
+ * Signature for EntityCrudRepository classes
+ *
+ * @typeParam E - An entity class
+ * @typeParam IdType - ID type for the entity
+ * @typeParam Relations - Relations for the entity
+ */
+export interface EntityCrudRepositoryClass<
+  E extends Entity,
+  IdType,
+  Relations extends object
+> extends CrudRepositoryClass<E, EntityCrudRepository<E, IdType, Relations>> {}
+
+/**
+ * Create (define) a CRUD repository class for the given model.
+ *
+ * @example
+ *
+ * ```ts
+ * const AddressRepository = defineCrudRepositoryClass(Address);
+ * ```
+ *
+ * @param modelClass - A model class such as `Address`.
+ *
+ * @typeParam M - Model class
+ * @typeParam R - Repository class/interface
+ */
+export function defineCrudRepositoryClass<
+  M extends Model,
+  R extends CrudRepository<M>
+>(
+  modelClass: typeof Model & {prototype: M},
+  baseRepositoryClass: Constructor<R>,
+): Constructor<R> {
```

Comment:
I think this should be `CrudRepositoryClass<M>` to communicate the fact that the class constructor accepts only a single `dataSource` argument?